Use conda integrated package management feature within snakemake

Dhananjhay opened the pull request:
Excited to report that I successfully ran an entire workflow of `hippunfold` **(for modality T1w)** on a test dataset (lowresMRI) using the `--use-conda` flag and without `--use-singularity`. I ran it once on the CBS server and another time on a vanilla `Ubuntu 24.04.1 LTS` Docker container.

To run the full end-to-end pipeline, I had to make some fundamental changes, some of which have their own dedicated PRs, and others need immediate review here:

- Since the Equivolume optoin uses `equivolumetric layering` method from the `nighres` package, which has proven difficult to integrate into the Conda ecosystem, I temporarily switched the default option of `--laminar_coords_method` in `snakebids.yaml` from `equivolume` to `laminar`. This change is still a work in progress (WIP).
- Laynii native #349 — We expect a robust solution soon!

Due to conflicting versions of libitk required by `ants v2.5.1` and `c3d v1.4.2`, I had to replace `antsApplyTransforms` with `greedy` in the one instance where they were used together. The newest version of `ants` available on conda-forge is `v2.5.4`, which has a compatible `libitk` version with `c3d`, but it currently segfaults during runtime. This issue is being actively investigated:

- [Segmentation fault (core dumped) error on using antsApplyTransforms tool from ants conda package](https://github.com/ANTsX/ANTs/issues/1825)
- [Segmentation fault in antsApplyTransforms using build h8c3795b_0 on Linux](https://github.com/conda-forge/ants-feedstock/issues/19)

The issue can be reproduced at [subfields.smk#L47](https://github.com/khanlab/hippunfold/blob/integrate-conda/hippunfold/workflow/rules/subfields.smk#L47).

**Update**
- Also worked for **T2w** modality (used thickSlice dataset)

akhanf commented on hippunfold/workflow/envs/env2.yaml:
can rename this env `c3d` or `convert3d`

akhanf commented on hippunfold/workflow/envs/env4.yaml:
can rename this one `ants`

akhanf commented on hippunfold/workflow/envs/env5.yaml:
can rename this one`greedyreg` or `greedy`

akhanf commented on hippunfold/workflow/envs/env6.yaml:
can rename this one `niftyreg`

akhanf commented on hippunfold/workflow/rules/preproc_t2.smk:
```diff
@@ -242,6 +248,8 @@ rule reg_t2_to_t1:
     container:
         config["singularity"]["autotop"]
+    conda:
+        "../envs/env7.yaml"
```
I am thinking we can split this into two rules so we don't have to make a "compound" environment that has niftyreg and c3d..  (so we can eliminate env7)

akhanf replied:
same with the other use of env7

Dhananjhay replied:
For the lack of creativity on my end, I've split and named the rules reg_t2_to_t1 and reg_t2_to_ref to reg_t2_to_t1_part1 and reg_t2_to_t1_part2, and reg_t2_to_ref_part1 and reg_t2_to_ref_part2, respectively.

akhanf replied:
I ended up renaming the rule to indicate it is for converting from ras to itk..

akhanf commented on hippunfold/workflow/envs/env8.yaml:
could call this one `pyvista`

akhanf commented on hippunfold/workflow/envs/env9.yaml:
maybe could call this something like `pydata`? 
chatgpt tells me `neurovis` is a good name too :)

akhanf commented on hippunfold/workflow/envs/env1.yaml:
maybe call this `pyunfold`?

akhanf commented on hippunfold/workflow/envs/env11.yaml:
```diff
@@ -0,0 +1,7 @@
+  - defaults
+dependencies:
+  - python=3.9
+  - nnunet=1.7.1
```
nnunet 1.7.1 works ok with CPU-based inference? Looks like you've tested it out already so that's great if so, we have been using that custom fork for awhile now would be good to get rid of it..

Dhananjhay replied:
Yes! I was skeptical at first and was ready to put `v1.6.6` on `khanlab` conda repo but  `v1.7.1` ended up working just fine; interestingly this was the only version available on conda-forge.

akhanf commented on hippunfold/workflow/rules/subfields.smk:
```diff
@@ -36,13 +36,15 @@ rule resample_unfoldreg_subfields:
     container:
         config["singularity"]["autotop"]
+    conda:
+        "../envs/env12.yaml"
```
I can't recall if this rule is actually used in the latest nativesurf PR, it may end up being obsolete, in which case we can hopefully drop env12 (if not then we should just figure out a cleaner way to do this..)

Dhananjhay replied:
I can see that this rule was preserved in the latest merge of `dev-v2.0.0` into this branch `integrate-conda` so we should figure out a cleaner way to do this. One alternate solution is to split this rule like before and get rid of environment `env12` altogether.

akhanf replied:
I haven't removed the rule, but looks like it is not actually used in the workflow anymore. I haven't gone through and pruned any unused rules yet -- that might identify a couple other rules that could be removed as well.

akhanf replied:
I've removed the unused rules (and subsequently unused envs). Since create_warps doesn't use naturalneighbor we can use the pyunfold env for that script.

akhanf commented on hippunfold/workflow/envs/env-python.yaml:
assuming this was for testing and will be dropped?

Dhananjhay replied:
Yes, it's been deleted now!

akhanf commented on hippunfold/workflow/envs/env11.yaml:
could rename `nnunet`

akhanf commented on hippunfold/workflow/envs/env10.yaml:
should try to use env1 for this (could perhaps do that in the latest PR where naturalneighbour is dropped)..

Dhananjhay replied:
Yes, I was going for only one env with all the python dependencies but that's not possible due to `numpy` version conflicts introduced by `naturalneighbor`.

akhanf commented:
This is amazing! Thanks for pushing this through.. 
I've made some comments on suggestions on what to name the various numbered envs for easier maintainance/development in the future, and where things could be merged/dropped too..

akhanf commented:
I've merged in #345 now, which means this will have some conflicts to resolve now.. I'll try to do that now (should be pretty straightforward, just alot of things moving around or being removed in that PR)..

akhanf commented:
I've removed the unused rules (and subsequently unused envs). Since create_warps doesn't use naturalneighbor we can use the pyunfold env for that script.

akhanf commented:
Got a chance to try this out with conda!

I installed miniconda, then installed snakemake (instructions from snakemake docs), and in that conda env then did `conda install snakebids` and then `conda install ruamel.yaml` (was missing from snakebids, I recall you have an open snakebids issue on something related?)

Then I just ran hippunfold with ./hippunfold/run.py directly (rather than pip installing as I didn't want to install any more dependencies, but I am guessing we can also make a hippunfold conda package (that depends on snakebids, snakemake). 

Was on my way, but then I realized that dev-v2.0.0 now depends on a new dev version of workbench (for the new hippocampus cifti integration), so for now we may just have to use the container for those rules (otherwise they will fail).. unless we want to make a temporary workbench conda package for the dev build?

akhanf commented:
> > I installed miniconda, then installed snakemake (instructions from snakemake docs), and in that conda env then did `conda install snakebids` and then `conda install ruamel.yaml` (was missing from snakebids, I recall you have an open snakebids issue on something related?)

> 
> Right, there's an open PR that resolves the issue with `ruamel.yaml`. It's ready to be merged—just needs your approval!
> 

I'll have a look at that now - if we can get the tests passing then can merge it in

> > Then I just ran hippunfold with ./hippunfold/run.py directly (rather than pip installing as I didn't want to install any more dependencies, but I am guessing we can also make a hippunfold conda package (that depends on snakebids, snakemake).
> 
> I just uploaded `snakebids v0.13.1` to Bioconda, and the required `snakemake` versions are already available. So, we're in good shape to publish `hippunfold v2.0.0` on Conda!
> 

Great - but no need to publish a release yet, unless you want to publish some test releases to iron out the process.. 

> > Was on my way, but then I realized that dev-v2.0.0 now depends on a new dev version of workbench (for the new hippocampus cifti integration), so for now we may just have to use the container for those rules (otherwise they will fail).. unless we want to make a temporary workbench conda package for the dev build?
> 
> The `--use-singularity` flag takes precedence over the `--use-conda` flag, so any rules with the `container` directive will default to using Singularity containers instead of Conda environments. I agree—we could temporarily host the new dev version of Workbench on our custom `khanlab` Conda repository until the package authors release it on Conda-Forge. Is the latest dev version available on Workbench's official GitHub repository? The only copy I could find seems to be hosted on [Dropbox](https://www.dropbox.com/scl/fi/k8nxg3f11xzmtv2msk8x2/workbench-linux64-dev_latest.zip?rlkey=1npndd9lc11i8q4nld2wz5o2x&dl=0)

Sure we could post the temporary one in our khanlab channel - it's not on their github as its not a release yet, it was passed along to us here https://balsa.wustl.edu/myelin and I put it on my dropbox for ease of download url.

Dhananjhay commented:
> Great - but no need to publish a release yet, unless you want to publish some test releases to iron out the process..

I've published a release based on  `integrate-conda` branch on the custom khanlab for the moment; don't forget to add channels `bioconda` and `conda-forge` in your local conda configuration before installing it. I was also able to run an entire workflow for modality T1w using this package 🥳 Once we release `v2.0.0` it'd take no time to publish an official release to the bioconda repo.

> Sure we could post the temporary one in our khanlab channel - it's not on their github as its not a release yet, it was passed along to us here https://balsa.wustl.edu/myelin and I put it on my dropbox for ease of download url.

In the latest commit, I've reverted back to the latest workbench labels and modified `workbench.yaml` to use `dev-connectome-workbench` package hosted on our khanlab conda repo. I tested these changes by running the following command:

```
hippunfold lowresMRI/ test-lowresMRI participant --participant-label 01 --modality T1w --use-conda --cores all --output_density 0p5mm 2mm unfoldiso --autotop-labels hipp
```

I'm planning on following up this with more extensive testing, more specifically by running the wet-test run script written by Jordan.

akhanf commented:
OK great - I'll merge upstream changes in and resolve some conflicts, but then I think we could merge this PR in?

Dhananjhay commented:
Yes, I think we are ready to merge!

akhanf commented:
actually, I'm going to merge in #359 then #349 first, so workflow doesn't rely on nighres..

akhanf commented:
ok got things up to date with upstream -- just need to add in a new laynii environment (I already refer to it as laynii.yaml), and re-run a wet test to make sure nothing is broken..

Dhananjhay commented:
I already had put laynii on our custom khanlab repo weeks ago so we don't have to wait to start testing!

akhanf commented:
everything here looks good so merging it in!
